### DISCUSSION TITLE
Could not resolve symbol

### DISCUSSION THREAD

**Discussion by try-agaaain** [bpftrace/bpftrace#2707] [Q&A] [answered]
I wrote a simple C language program named "test1," and I used the command `sudo bpftrace -l 'uprobe:test1:*` to list all the available probe points as follows:

```shell
$ sudo bpftrace -l 'uprobe:test1:*' -v
WARNING: No DWARF found for "test1", cannot show parameter info
uprobe:test1:__do_global_dtors_aux
uprobe:test1:_fini
uprobe:test1:_init
uprobe:test1:_start
uprobe:test1:deregister_tm_clones
uprobe:test1:frame_dummy
uprobe:test1:main
uprobe:test1:register_tm_clones
```

Next, I wanted to monitor the "main" function and wrote the corresponding bpftrace command, but encountered an error:

```shell
$ sudo bpftrace -e 'uprobe:test1:main /comm=="test1"/ {printf("info: %d", pid);}'
Attaching 1 probe...
ERROR: Could not resolve symbol: test1:main
```

I also tried monitoring other functions, but encountered errors for each of them:

```shell
sudo bpftrace -e 'uprobe:test1:frame_dummy /comm=="test1"/ {printf("info: %d", pid);}'
Attaching 1 probe...
ERROR: Could not determine boundary for frame_dummy (symbol has size 0). Use --unsafe to force attachment.

$ sudo bpftrace -e 'uprobe:test1:frame_dummy /comm=="test1"/ {printf("info: %d", pid);}' --unsafe
Attaching 1 probe...
ERROR: Could not resolve symbol: test1:frame_dummy
```

I wonder why these functions cannot be monitored.

**Comment by viktormalik** [comment, marked as answer]
The second part of a uprobe specification must be a *path* to an executable. That is, if `test1` is in the current directory, your probes should start with `uprobe:./test1:<function>`.

**Reply by try-agaaain** [reply]
Thank you very much. And I want to trace the `open` function in `libc` and print its arguments, I learn about that the function declaration of `open` is `int open(const char *__file, int __oflag, ...)`. Therefore, I wrote the following bpftrace program, but encountered an error:

```shell
sudo bpftrace -e 'uprobe:/lib/x86_64-linux-gnu/libc.so.6:open {printf("pid:%d, __file:%s\n", pid, args->__file);}'
stdin:1:1-44: ERROR: No debuginfo found for /lib/x86_64-linux-gnu/libc.so.6
uprobe:/lib/x86_64-linux-gnu/libc.so.6:open {printf("pid:%d, __file:%s\n", pid, args->__file);}
```

Is there any way to properly retrieve and print the output of the `open` function?

**Reply by viktormalik** [reply]
You need DWARF (debugging information) for the library installed in order to access function arguments by name via `args`. So, either install it (should be something like `apt install libc6-dbg` on Ubuntu) or directly access the first argument using `arg0`.

Hint: you don't need to specify full path with libraries, `uprobe:libc:open` should do the job.

**Reply by try-agaaain** [reply]
I tried to use `arg0` to get the parameters, but it only gives me the number representing the file handle:

```shell
$ sudo bpftrace -e 'uprobe:libc:open {printf("pid:%d, filename:%d\n", pid, arg0);}'
Attaching 1 probe...
pid:509, filename:110633776
pid:509, filename:116932960
pid:509, filename:107637648
pid:509, filename:116932960
...
```

Is there a way to get the corresponding file names?

**Reply by viktormalik** [reply]
`arg0` is a pointer and you need to convert it to string with the `str` builtin: `printf("filename: %s\n", str(arg0))`.

**Reply by try-agaaain** [reply]
It works well. Thanks a lot.
